### PR TITLE
refactor: update report system design

### PR DESCRIPTION
**Pull request by joseluis-teran-jalafund**
## Trello Ticket

**[Task - 07](https://trello.com/c/yCBD1yFd/7-task-7-reporting)**

## Description

This **PR** is on charge of updating the UML for the **Report System** provided. For now the refactor renames and updates the contract and includes an only base concrete for report objects.

## Type of Change

Put an `x` in all the boxes that apply:

- [ ] ✨ New Feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug Fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking Change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code Refactor
- [ ] ✅ Build Configuration Change
- [x] 📝 Documentation
- [ ] 🗑️ Chore

## Pull Request Checklist

Put an `x` in all the boxes that apply:

- [x] My commit messages are detailed.
- [x] My code follows the code style of this project.
- [ ] No existing features have been broken without good reason.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: UML-Diagram/ReportSystem.puml [hunk 1]
```diff
@@ -1,32 +1,28 @@
...
+    ' Define main abstract base
+    abstract class BaseReport implements IReport {
+        - BorrowManager : _borrowManager
     }
 
     ' Define main classes
-    class BookCurrentBorrwedReport extends ABorrowReport {
-        + BookBorrowedReportFormat()
+    class BorrowPatronHistory extends BaseReport {
+        + PatronBorrowHistoryReport() : void
```
Comment (alex-paca-jalafund) [inline UML-Diagram/ReportSystem.puml:16]: I think that its missing a client class wich you can call and set the kind of BorrowReport and an execute method to  display the report, but whay you think

**Review by alex-paca-jalafund — CHANGES_REQUESTED** [review]

File: UML-Diagram/ReportSystem.puml (hunk 1 above)
Comment (joseluis-teran-jalafund) [reply UML-Diagram/ReportSystem.puml:16]: We can create a new class as you suggest but it won't be the type of client. It should be a concrete for borrows client classes.
But if we review our actual code on **develop** branch, `BookReport` and `PatronReport` classes, first of all implements `Reporting` interface, which only has the next function.
```
interface Reporting
{
    public void generateReport();
}
```
So, `BookReport` and `PatronReport` classes have the next code.
```
public abstract class BookReport: Reporting
{
    protected BorrowManager BorrowList;
    public BookReport( BorrowManager BorrowList)
    {
        this.BorrowList = BorrowList;
    }

    public void generateReport()
    {
        Console.WriteLine("This is a book borrow Report");
    }
}
```
```
public abstract class PatronReport : Reporting
{
    protected BorrowManager Borrows;
    public PatronReport(BorrowManager Borrows)
    {
        this.Borrows = Borrows;
    }

    public void generateReport()
    {
        Console.WriteLine("This is a Patron borow Report");
    }
}
```
As result, our client classes will have the next code. Where concrete class share its attribute and let client class access the `generateReport` method and its content, which comes from the contract.
```
public class BorrowPatronHistory : PatronReport
{
    public BorrowPatronHistory(BorrowManager BorrowList) : base(BorrowList){}

    public void PatronBorrowHistoryReport(string PatronName)
    {
        generateReport();
        foreach(Borrow borrow in Borrows.GetBorrows()){
            if(borrow.GetPatron().getName() == PatronName){
                borrow.BorrowDetails();
            }
        }        
    }
}
```
To sum up, if we're still creating concretes for specific client classes, we'll have one for everyone. Having as result redundancy classes and violating **DRY** principle instead of making of use a general or base concrete class for client classes.

Comment (joseluis-teran-jalafund) [reply UML-Diagram/ReportSystem.puml:16]: We can create a new class as you suggest but it won't be the type of client. It should be a concrete for borrows client classes.
But if we review our actual code on **develop** branch, `BookReport` and `PatronReport` classes, first of all implements `Reporting` interface, which only has the next function.
```
interface Reporting
{
    public void generateReport();
}
```
So, `BookReport` and `PatronReport` classes have the next code.
```
public abstract class BookReport: Reporting
{
    protected BorrowManager BorrowList;
    public BookReport( BorrowManager BorrowList)
    {
        this.BorrowList = BorrowList;
    }

    public void generateReport()
    {
        Console.WriteLine("This is a book borrow Report");
    }
}
```
```
public abstract class PatronReport : Reporting
{
    protected BorrowManager Borrows;
    public PatronReport(BorrowManager Borrows)
    {
        this.Borrows = Borrows;
    }

    public void generateReport()
    {
        Console.WriteLine("This is a Patron borow Report");
    }
}
```
As result, our client classes will have the next code. Where concrete class share its attribute and let client class access the `generateReport` method and its content, which comes from the contract.
```
public class BorrowPatronHistory : PatronReport
{
    public BorrowPatronHistory(BorrowManager BorrowList) : base(BorrowList){}

    public void PatronBorrowHistoryReport(string PatronName)
    {
        generateReport();
        foreach(Borrow borrow in Borrows.GetBorrows()){
            if(borrow.GetPatron().getName() == PatronName){
                borrow.BorrowDetails();
            }
        }        
    }
}
```
To sum up, if we're still creating concretes for specific client classes, we'll have one for everyone. Having as result redundancy classes and violating **DRY** principle. That's why I suggest to use a general or base concrete class for client classes.

Comment (alex-paca-jalafund) [reply UML-Diagram/ReportSystem.puml:16]: That are the concrete classes, that are not the client im refering, BorrowPatronHistory its a conrete of PatronReport, and  i agree for merging both templates in one because has the same structure the PatronReport and BookReport, but again it will be usefull to have a client class that uses the interface has a paremeter in order to set new concretes of the new ReportTemplate and doesnt has to initialize every class. Use the reference to the family/group of methods the interface in order to only instance one class that can set the different types of concretes that are using the interface. Something like:

public class ReportClient
+ SetReport (IReport report) : void -> in this method set the type of Report
+ ExecuteReport () : void -> this method execute the method of the seted Report

Comment (joseluis-teran-jalafund) [reply UML-Diagram/ReportSystem.puml:16]: What do you mean by "initialize every class"? Are you talking about `BorrowManager` instance of `BaseReport` class?

Comment (joseluis-teran-jalafund) [reply UML-Diagram/ReportSystem.puml:16]: Are you we really need another class to do that? With the use of IReport in some class, we already can do that. So I don't think we need one class to use the concretes if we can do it with the contract.

Comment (alex-paca-jalafund) [reply UML-Diagram/ReportSystem.puml:16]: It can be, it will work, but that doesnt reflect in the UML, IReport doesnt have a relation from his use in other class

File: UML-Diagram/ReportSystem.puml [hunk 2]
```diff
@@ -1,32 +1,28 @@
...
-    class OverdueBooksReport extends ABorrowReport {
-        + OverdueBooksFormat()
+    class CurrentBorrowBooksReport extends BaseReport {
+        + BorrowedBooksReport() : void
     }
 
-    class BorrowingHistoryReport extends APatronReport {
-        + BorrowHistoryFormat()
+    class OverdueBooksReport extends BaseReport {
+        + OverdueBooksListReport() : void
```
Comment (alex-paca-jalafund) [inline UML-Diagram/ReportSystem.puml:24]: Name of methods its different from the interface

**Review by alex-paca-jalafund — CHANGES_REQUESTED** [review]

File: UML-Diagram/Master-UML.puml [hunk 3]
```diff
@@ -48,8 +49,6 @@ package "Library system"{
     StaffMenu ..> ReportStaffMenu
     StaffMenu ..> StaffSearchMenu
     Login ..> StaffMenu
-    ABorrowReport ..* BorrowManager
-    APatronReport ..* PatronManager
```
Comment (alex-paca-jalafund) [inline UML-Diagram/Master-UML.puml:52]: The use of the IReport has we are talking in the conversation it should reference StaffReportMenu, i dont remember if its in use in other classes

**Review by alex-paca-jalafund — CHANGES_REQUESTED** [review]

File: UML-Diagram/Master-UML.puml (hunk 3 above)
Comment (joseluis-teran-jalafund) [reply UML-Diagram/Master-UML.puml:52]: Now it's with its relationships but there is no a class according to the actual code into **develop** branch, which is `BorrowPatronHistoryInput`. I guess we can add it later, right?

File: UML-Diagram/ReportSystem.puml (hunk 2 above)
Comment (joseluis-teran-jalafund) [reply UML-Diagram/ReportSystem.puml:24]: Can you review the code we have on **develop** branch to verify please?

For example, `OverdueBooksReport` class has this code, which I represented on the UML and interface method it's inside of it due to we're using for printing a message in the console. That's why those classes have those method names.
```
namespace ReportSystem.Concrete;
public class OverdueBooksReport : BookReport
{
    public OverdueBooksReport(BorrowManager BorrowList) : base(BorrowList){}

    public void OverdueBooksListReport()
    {
        generateReport();
        foreach(Borrow borrow in BorrowList.GetBorrows()){
            if(DateTime.Now > borrow.GetDueDate()){
                borrow.BorrowDetails();
            }
        }        
    }
}
```

Comment (alex-paca-jalafund) [reply UML-Diagram/ReportSystem.puml:24]: i reviewed and yeah can be managed that way

File: UML-Diagram/Master-UML.puml (hunk 3 above)
Comment (alex-paca-jalafund) [reply UML-Diagram/Master-UML.puml:52]: Maybe if its needed an update but as i see the only use for Report was there on ReportStaffMenu so yes.

**Review by alex-paca-jalafund — APPROVED** [review]
Nice
